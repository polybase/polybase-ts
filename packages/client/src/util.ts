import { CollectionRecord } from './Record'
import type { CallArg, FieldTypes } from './types'
import { Root as AST, Property as ASTProperty, Collection as ASTCollection, CollectionAttribute, ObjectField } from '@polybase/polylang/dist/ast'

export function getCollectionASTFromId(id: string, ast: AST): ASTCollection | undefined {
  return getCollectionASTFromName(getCollectionShortNameFromId(id), ast)
}

export function getCollectionASTFromName(name: string, ast: AST): ASTCollection | undefined {
  const collections = ast.filter((n) => n.kind === 'collection') as ASTCollection[]
  return collections.find((n: ASTCollection) => n.name === name)
}

export function getCollectionProperties(collection: ASTCollection): ASTProperty[] {
  return collection.attributes.filter((a: CollectionAttribute) => a.kind === 'property') as ASTProperty[]
}

export function getCollectionShortNameFromId(id: string): string {
  const name = id.split('/').pop()
  if (!name) throw new Error(`Invalid collection id: ${id}`)
  return name
}

export function encodeBase64(value: Uint8Array): string {
  return btoa(String.fromCharCode.apply(null, value as unknown as number[]))
}

export function decodeBase64(value: string): Uint8Array {
  const binaryString = atob(value)

  const bytes = new Uint8Array(binaryString.length)
  for (let i = 0; i < binaryString.length; i++) {
    bytes[i] = binaryString.charCodeAt(i)
  }

  return bytes
}

export function referenceArg(arg: CallArg): CallArg {
  if (arg instanceof CollectionRecord) return arg.reference()

  if (Array.isArray(arg)) {
    for (const i in arg) {
      arg[i] = referenceArg(arg[i]) as FieldTypes
    }
  }

  return arg
}

export function serializeValue(arg: CallArg): CallArg {
  if (arg instanceof Uint8Array) return encodeBase64(arg)
  arg = referenceArg(arg)
  return arg
}

export function deserializeRecord(data: Record<string, any>, properties: (ASTProperty | ObjectField)[]) {
  if (!data) return

  for (const property of properties) {
    switch (property.type.kind) {
      case 'primitive':
        switch (property.type.value) {
          case 'bytes':
            if (property.name in data) {
              data[property.name] = decodeBase64(data[property.name])
            }
        }
        break
      case 'object':
        deserializeRecord(data[property.name], property.type.fields)
        break
    }
  }
}

// Adds a key/value to a record, or creates an object with the key/value
export function addKeyValue(key: string, value: any, obj?: Record<string, any>): Record<string, any> {
  const o = obj ?? {}
  o[key] = value
  return o
}

// Removes given keys from the object
export function removeKey(key: string, obj?: any): Record<string, any> {
  if (!obj || !isPlainObject(obj)) return {}
  for (const k of key) {
    if (k in obj) {
      delete obj[k]
    }
  }
  return obj
}

// Returns true if the object is a plain object
export function isPlainObject(val: any): val is Record<string, any> {
  return typeof val === 'object' && val.constructor === Object
}
