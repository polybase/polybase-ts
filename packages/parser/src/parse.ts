import nearley from 'nearley'
/* eslint-disable-next-line import/no-unresolved, import/extensions */
import grammer from './lexer'

const { Parser } = nearley as any

Parser.prototype.reportError = function reportError (token: any) {
  const lines = []
  const tokenDisplay = (token.type ? `${token.type} token: ` : token) + JSON.stringify(token.value !== undefined ? token.value : token)
  lines.push(this.lexer.formatError(token, 'Syntax error'))
  lines.push(`Unexpected ${tokenDisplay}. Instead, I was expecting to see one of the following:\n`)
  // Removed lines, because they take a very long time to render!
  lines.push('')
  return lines.join('\n')
}

export default function parse (formula: string) {
  const parser = new nearley.Parser(nearley.Grammar.fromCompiled(grammer))
  return parser.feed(formula).results[0]
}
