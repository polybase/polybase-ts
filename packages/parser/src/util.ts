
import map from 'lodash/map'
import flatMap from 'lodash/flatMap'
import filter from 'lodash/filter'
import isArray from 'lodash/isArray'
import isObject from 'lodash/isObject'
import mapValues from 'lodash/mapValues'
import { Node } from './types'

export function filterNodes (node: Node, filterFn: (n: Node) => boolean): Node[] {
  if (!node) return []
  if (isArray(node)) return flatMap(node, n => filterNodes(n, filterFn))
  const {
    left, op, right, value, args, prefix,
  } = node as any
  const nodes: Node[] = flatMap(
    filter([left, op, right, value, args, prefix], val => val !== undefined),
    n => filterNodes(n, filterFn),
  )
  if (filterFn(node)) nodes.push(node)
  return nodes
}

const nodeProps = ['left', 'op', 'right', 'value', 'args', 'prefix', 'body', 'params', 'expression']
export function mapNodes (node: Node|Node[], mapFn: (n: Node) => Node): Node|Node[] {
  if (isArray(node)) return map(node, n => mapNodes(n, mapFn)) as Node[]
  const mapped = mapFn(node)
  return mapValues(mapped, (val, key) => {
    if (nodeProps.indexOf(key) > -1) {
      if (isArray(val)) return map(val, n => mapNodes(n, mapFn))
      if (isObject(val)) return mapNodes(val, mapFn)
    }
    return val
  }) as Node
}
