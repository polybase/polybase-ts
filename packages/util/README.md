# @[polybase](https://polybase.xyz)/util

A utility library to help support common [Polybase](https://polybase.xyz) utility functions.

## Browser Tests

We use `puppeteer` to run our tests in the browser.
We also run the same tests in Node.js.

Only the function in the first argument of `testNodeAndBrowser` runs in the browser (and Node).
This function can't depend on any code outside of the function.
It might also fail if given code that depends on features from ES version higher than
the ES version specified in ts-jest's tsconfig, as it will depend on extra Typescript runtime code.
It's best to keep the function as simple as possible.
