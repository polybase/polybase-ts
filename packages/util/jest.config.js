const tsPreset = require('ts-jest/jest-preset')
// const puppeteerPreset = require('jest-puppeteer/jest-preset')

const config = {
  verbose: true,
  // ...puppeteerPreset,
  ...tsPreset,
  roots: [
    '<rootDir>/src',
  ],
  testMatch: [
    '<rootDir>/**/*.spec.{js,jsx,ts,tsx}',
  ],
  testEnvironment: 'node',
  globals: {
    test_url: `http://${process.env.HOST || '127.0.0.1'}:${process.env.PORT || 3000}`,
  },
}

module.exports = config
