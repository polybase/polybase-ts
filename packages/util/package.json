{
  "name": "@spacetimexyz/util",
  "description": "Utility functions for Spacetime",
  "version": "0.1.27",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "license": "MIT",
  "files": [
    "dist/*"
  ],
  "scripts": {
    "clean": "rimraf dist/* &",
    "build": "yarn clean && tsc",
    "prepare": "yarn build",
    "test": "jest ./src",
    "fix": "yarn eslint \"./src/**/*.{js,jsx}\" --fix"
  },
  "jest": {
    "preset": "ts-jest",
    "roots": [
      "<rootDir>/src"
    ],
    "testMatch": [
      "<rootDir>/**/*.spec.{js,jsx,ts,tsx}"
    ],
    "testEnvironment": "node"
  },
  "devDependencies": {
    "@sinonjs/fake-timers": "^9.1.2",
    "@types/jest": "^28.1.7",
    "@types/sinonjs__fake-timers": "^8.1.2",
    "@typescript-eslint/eslint-plugin": "^5.25.0",
    "@typescript-eslint/parser": "^5.25.0",
    "eslint": "^8.22.0",
    "eslint-config-standard": "^17.0.0",
    "eslint-plugin-import": "^2.26.0",
    "eslint-plugin-jest": "^26.8.7",
    "eslint-plugin-n": "^15.2.0",
    "eslint-plugin-promise": "^6.0.0",
    "jest": "^28.1.3",
    "rimraf": "^3.0.2",
    "ts-jest": "^28.0.8",
    "typescript": "^4.6.4"
  },
  "dependencies": {
    "@ethersproject/bytes": "^5.7.0",
    "@ethersproject/signing-key": "^5.7.0",
    "tweetnacl": "^1.0.3",
    "tweetnacl-util": "^0.15.1"
  }
}
