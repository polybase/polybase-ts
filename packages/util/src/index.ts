export * from './util'
export * from './deprecated'
export { BytesLike } from '@ethersproject/bytes'
export * as aescbc from './algorithems/aes-cbc'
export * as secp256k1 from './algorithems/secp256k1'
export * as x25519xsalsa20poly1305 from './algorithems/x25519-xsalsa20-poly1305'
export * from './types'
