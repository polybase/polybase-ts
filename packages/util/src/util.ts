import { arrayify, hexlify } from '@ethersproject/bytes'
import { encodeBase64, decodeBase64, encodeUTF8, decodeUTF8 } from 'tweetnacl-util'

export function stringifiableToHex (value: any) {
  return hexlify(Buffer.from(JSON.stringify(value)))
}

export function isNullish (value: any) {
  return value === null || value === undefined
}

export function stripPublicKeyPrefix (publicKey: Uint8Array): Uint8Array {
  if (publicKey.byteLength % 32 === 0) return publicKey
  return publicKey.slice(1)
}

export function addPublicKeyPrefix (publicKey: Uint8Array): Uint8Array {
  if (publicKey.byteLength === 64) return Buffer.concat([Buffer.from([0x4]), publicKey])
  return publicKey
}

export function encodeToString (data: Uint8Array, encoding: 'hex'|'base64'|'utf8'): string {
  if (encoding === 'hex') return hexlify(data)
  if (encoding === 'utf8') return encodeUTF8(data)
  if (encoding === 'base64') return encodeBase64(data)
  throw new Error('Invalid encoding, must be either hex, utf8 or base64')
}

export function decodeFromString (data: string, encoding: 'hex'|'base64'|'utf8'): Uint8Array {
  if (encoding === 'hex') {
    if (!data.startsWith('0x')) {
      return arrayify(`0x${data}`)
    }
    return arrayify(data)
  }
  if (encoding === 'utf8') return decodeUTF8(data)
  if (encoding === 'base64') return decodeBase64(data)
  throw new Error('Invalid encoding, must be either hex, utf8 or base64')
}

export function stringifyEncryptedData (data: any, encoding: 'hex'|'base64' = 'hex'): string {
  const obj: any = {}
  Object.keys(data).forEach((key: any) => {
    obj[key] = typeof data[key] === 'string' ? data[key] : encodeToString(data[key] as Uint8Array, encoding)
  })
  const str = JSON.stringify(obj)
  const buf = decodeFromString(str, 'utf8')
  return encodeToString(buf, encoding)
}

export function parseEncrypedData (str: string, encoding: 'hex'|'base64' = 'hex'): any {
  const buf = decodeFromString(str, encoding)
  const data = JSON.parse(encodeToString(buf, 'utf8'))
  const obj: any = {}
  Object.keys(data).forEach((key: any) => {
    if (key === 'version') obj[key] = data[key]
    else obj[key] = decodeFromString(data[key] as string, encoding)
  })
  return obj
}

export function assert (condition: boolean, message: string) {
  if (!condition) {
    throw new Error(message || 'Assertion failed')
  }
}
