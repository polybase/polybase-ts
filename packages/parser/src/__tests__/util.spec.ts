import { mapNodes } from '../util'
import parse from '../parse'

const exp = (str: string) => `function Hello () {\nreturn ${str}\n}`

describe('util', () => {
  describe('Map Nodes', () => {
    test('Map number without mod', () => {
      const ast = parse(exp('10'))
      expect(mapNodes(ast, node => ({ ...node }))).toEqual(ast)
    })

    test('Map complex without mod', () => {
      const ast = parse(exp('(10 * 10) + "hello" / func(name)'))
      expect(mapNodes(ast, node => ({ ...node }))).toEqual(ast)
    })

    test('Map and change number value', () => {
      const ast = parse(exp('10'))
      expect(mapNodes(ast, (node) => {
        if (node.type === 'number') return { ...node, text: '11', value: '11' }
        return node
      })).toEqual(parse(exp('11')))
    })

    test('Map and change function name value', () => {
      const ast = parse(exp('10 / func(name)'))
      expect(mapNodes(ast, (node: any) => {
        if (node.type !== 'function') return node
        return { ...node, prefix: { ...node.prefix, value: 'altFunc', text: 'altFunc' } }
      })).toMatchSnapshot()
    })
  })
})
