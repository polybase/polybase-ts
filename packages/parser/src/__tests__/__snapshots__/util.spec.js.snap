// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util Map Nodes Map and change function name value 1`] = `
Array [
  Object {
    "body": Array [
      Object {
        "expression": Object {
          "left": Object {
            "col": 8,
            "line": 2,
            "lineBreaks": 0,
            "offset": 27,
            "text": "10",
            "toString": [Function],
            "type": "number",
            "value": "10",
          },
          "op": Object {
            "col": 11,
            "line": 2,
            "lineBreaks": 0,
            "offset": 30,
            "text": "/",
            "toString": [Function],
            "type": "divide",
            "value": "/",
          },
          "right": Object {
            "args": Array [
              Object {
                "col": 18,
                "line": 2,
                "lineBreaks": 0,
                "offset": 37,
                "text": "name",
                "toString": [Function],
                "type": "name",
                "value": "name",
              },
            ],
            "prefix": Object {
              "col": 13,
              "line": 2,
              "lineBreaks": 0,
              "offset": 32,
              "text": "func",
              "toString": [Function],
              "type": "name",
              "value": "func",
            },
            "type": "call",
          },
          "type": "product",
        },
        "type": "return",
      },
    ],
    "name": Object {
      "col": 10,
      "line": 1,
      "lineBreaks": 0,
      "offset": 9,
      "text": "Hello",
      "toString": [Function],
      "type": "name",
      "value": "Hello",
    },
    "params": Array [],
    "prefix": Object {
      "text": "altFunc",
      "value": "altFunc",
    },
    "type": "function",
  },
]
`;
